active/0000-engines.md

```diff
@@ -0,0 +1,411 @@
+}
+```
+
+As a result, the engine's `application` template will be rendered into the
+`sidebar` outlet in the application's `main` template.
+
+### Loading phases
+
+Engines can exist in several phases:
+
+* Booted - an engine that's been installed in a parent application will have
+  its dependencies loaded and its (non-instance) initializers invoked when the
+  parent application boots.
+
+* Mounted - Routable and route-less engines have slightly different concepts of
+  "mounting". A routable engine is considered mounted when it has been included
+  by a router at one or more mount-points. A route-less engine is considered
+  mounted as soon as a route's `mount` call resolves.
+
+* Instantiated - When an engine is instantiated, an `EngineInstance` is created
+  and an engine's instance initializers are invoked. A routable engine is
+  instantiated when a route is visited at or beyond its mount-point. A
+  route-less engine is instantiated as soon as it is mounted.
+
+Special `before` and `after` hooks could be added to application instance
+initializers that allow them to be ordered relative to engine instance
+initializers.
+
+### Engine boundaries
+
+Besides its routing map, an engine does not share any other resources with its
+parent by default. Engines maintain their own registries and containers, which
+ensure that they stay isolated. However, some explicit sharing of resources
+between engines and parents is allowed.
+
+#### Engine / parent dependencies
+
+Dependencies between engines and parents can be defined imperatively or
+declaratively.
+
+Imperative dependencies can be defined in an engine's instance initializers.
+When an engine is instantiated, the `parent` property on its `EngineInstance` is
+set to its parent instance (either an `ApplicationInstance` or
+`EngineInstance`). Since the engine instance is available in the instance
+initializer, this `parent` property can also be accessed. This allows an engine
+instance to interrogate its parent, specifically through its `RegistryProxy` and
+`ContainerProxy` interfaces.
+
+Alternatively, declarative dependencies can be defined on a limited basis. The
+initial API will be limited: an engine can define an array of `services` that it
+requires from its parent.
+
+For example, the following engine expects its parent to provide `store` and
+`session` services:
+
+```
+import Ember from 'ember';
+
+var Engine = Ember.Engine.extend({
+  dependencies: {
+    services: [
+      'store',
+      'session'
+    ]
+  }
+});
+
+export default Engine;
+```
+
+The parent application can provide a re-mapping of services from its namespace
+to that of the engine via an `engines` declaration.
+
+In the following example, the application shares its `store` service directly
+with the `checkout` engine. It also shares its `current-user` service as the
+`session` service requested by the engine.
+
+```
+import Ember from 'ember';
+
+var App = Ember.Application.extend({
+  engines: {
+    checkout: {
```

Comment:
we can likely insert "some sort of DSL" to be derived, when we fully flesh out what it needs to accomplish.